File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2021/Day2.kt
package com.github.ephemient.aoc2021

/** [Day 2](https://adventofcode.com/2021/day/2): Dive! */
class Day2(private val lines: List<String>) {
    fun part1(): Int {
        var x = 0
        var y = 0
        for (line in lines) {
            val (prefix, suffix) = line.split(" ", limit = 2)
            when (prefix) {
                "forward" -> x += suffix.toInt()
                "down" -> y += suffix.toInt()
                "up" -> y -= suffix.toInt()
            }
        }
        return x * y
    }

    fun part2(): Int {
        var x = 0
        var y = 0
        var depth = 0
        for (line in lines) {
            val (prefix, suffix) = line.split(" ", limit = 2)
            when (prefix) {
                "forward" -> suffix.toInt().let {
                    x += it
                    y += it * depth
                }
                "down" -> depth += suffix.toInt()
                "up" -> depth -= suffix.toInt()
            }
        }
        return x * y
    }
}

File: py/aoc2021/day2.py
def part1(lines):
    """
    >>> part1(["forward 5", "down 5", "forward 8", "up 3", "down 8", "forward 2"])
    150
    """
    x, y = 0, 0
    for line in lines:
        match line.partition(" "):
            case ("forward", _, d):
                x += int(d)
            case ("down", _, d):
                y += int(d)
            case ("up", _, d):
                y -= int(d)
    return x * y


def part2(lines):
    """
    >>> part2(["forward 5", "down 5", "forward 8", "up 3", "down 8", "forward 2"])
    900
    """
    x, y, depth = 0, 0, 0
    for line in lines:
        match line.partition(" "):
            case ("forward", _, d):
                d = int(d)
                x += d
                y += d * depth
            case ("down", _, d):
                depth += int(d)
            case ("up", _, d):
                depth -= int(d)
    return x * y


parts = (part1, part2)
